hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSEditLog.java

```diff
@@ -1509,13 +1509,18 @@ synchronized void abortCurrentLogSegment() {
    * effect.
    */
   @Override
-  public synchronized void purgeLogsOlderThan(final long minTxIdToKeep) {
+  public synchronized void purgeLogsOlderThan(long minTxIdToKeep) {
     // Should not purge logs unless they are open for write.
     // This prevents the SBN from purging logs on shared storage, for example.
     if (!isOpenForWrite()) {
       return;
     }
-    
+
+    // Reset purgeLogsFrom to avoid purging edit log which is in progress.
+    if (isSegmentOpen()) {
+      minTxIdToKeep = minTxIdToKeep > curSegmentTxId ? curSegmentTxId : minTxIdToKeep;
```

Comment:
Just below this we have:
```java
    assert curSegmentTxId == HdfsServerConstants.INVALID_TXID || // on format this is no-op
      minTxIdToKeep <= curSegmentTxId :
      "cannot purge logs older than txid " + minTxIdToKeep +
      " when current segment starts at " + curSegmentTxId;
```
We assert that `minTxIdToKeep <= curSegmentTxId`. So in the situation you described where `minTxIdToKeep > curSegmentTxId`, shouldn't the assert fail?

Reply:
> Just below this we have:
> 
> ```java
>     assert curSegmentTxId == HdfsServerConstants.INVALID_TXID || // on format this is no-op
>       minTxIdToKeep <= curSegmentTxId :
>       "cannot purge logs older than txid " + minTxIdToKeep +
>       " when current segment starts at " + curSegmentTxId;
> ```
> 
> We assert that `minTxIdToKeep <= curSegmentTxId`. So in the situation you described where `minTxIdToKeep > curSegmentTxId`, shouldn't the assert fail?

Thank you @xkrogen very much for your reply. 

We did add assertion. But in the production, assertions are usually prohibited and do not actually take effect. I think we should add strict logic to avoid problems. What do you think of this?

Reply:
Hi @sunchao @tasanuma , could you please take a look at this discussion.

Reply:
This problem begin from inprogress edits tail. And this issue [HDFS-14317](https://issues.apache.org/jira/browse/HDFS-14317) does a good job of avoiding this problem.

However, if SNN's rolledit operation is disabled accidentally by configuration, and ANN's automatic roll period is very long, then edit log which is in progress may also be purged.

Although we add assertions, assertion is generally disabled in a production. This bug also proves that we are not strictly ensure`(inTxIdToKeep <= curSegmentTxId)`. So it is dangerous for NameNode. 

We should reset `minTxIdToKeep` to ensure that the in progress edit log is not purged very strict. And wait for ANN to automatically roll to finalize the edit log. Then, after checkpoint, ANN automatically purged the finalized editlog(See the stack mentioned above).

Reply:
Hi @virajjasani , please also take a look. Thanks a lot.